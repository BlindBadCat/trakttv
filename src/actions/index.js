import fetch from 'cross-fetch';
import C from '../constants';

const traktTVClientID = '127b74187bd5bc054477b5180794f87b02aa785b0a17e0476406a547fd1bda64';
const fanartTVClientID = 'a724beb35e84d4ff20154856b6ef2a99';

const headersForTraktTV = {
  'Content-Type': 'application/json',
  'trakt-api-version': '2',
  'trakt-api-key': traktTVClientID,
};

const traktTVfetchInit = {
  method: 'GET',
  headers: headersForTraktTV,
  Origin: 'localhost:3000',
  mode: 'cors',
};


const getPaginationFromFetchHeaders = headers => ({
  currentPage: parseInt(headers['x-pagination-page'], 10),
  pageCount: parseInt(headers['x-pagination-page-count'], 10),
  limit: parseInt(headers['x-pagination-limit'], 10),
  itemCount: parseInt(headers['x-pagination-item-count'], 10),
});

export const addShows = (shows, urlParams) => ({
  type: C.ADD_SHOWS,
  payload: {
    shows,
    urlParams,
  },
});

const shouldFetchMovies = state => !state.shows.isFetching;

const getShowsRequest = () => ({
  type: C.GET_SHOWS_REQUEST,
});

const fetchShowsError = error => ({
  type: C.GET_SHOWS_ERROR,
  payload: error,
});

export function resetForNewFetch() {
  return {
    type: C.RESET_SHOWS_FOR_PAGINATION,
  };
}

export const fetchShows = urlParams => (dispatch) => {
  const {
    searchUrl, pagination, query, genre,
  } = urlParams;
  const { currentPage, limit } = pagination;
  const urlForTraktTVFetch = `https://api.trakt.tv/${searchUrl}?extended=full&page=${currentPage}&limit=${limit}${query}${genre}`;
  dispatch(getShowsRequest());
  let newPagination;
  return fetch(urlForTraktTVFetch, traktTVfetchInit)
    .then((response) => {
      newPagination = getPaginationFromFetchHeaders(response.headers.map);
      return response.json();
    })
    .catch(error => dispatch(fetchShowsError(error)))
    .then(json => dispatch(addShows(json, { ...urlParams, pagination: { ...newPagination } })));
};
export function resetPostersForPagination() {
  return {
    type: C.RESET_FOR_NEW_FETCH,
  };
}


export function fetchShowsIfNeeded(urlParams) {
  return (dispatch, getState) => {
    if (shouldFetchMovies(getState())) {
      return dispatch(fetchShows(urlParams));
    }
  };
}

const shouldFetchPoster = state => !state.poster.isFetching;

const getPosterRequest = () => ({
  type: C.GET_POSTER_URL_REQUEST,
});

const getPosterUrl = (url, id) => ({
  type: C.GET_POSTER_URL_ADN_ID,
  payload: { id, url },
});

const fetchPoster = ids => (dispatch) => {
  dispatch(getPosterRequest());
  const len = ids.length;
  for (let i = 0; i < len; i += 1) {
    const fanartShowURL = `https://webservice.fanart.tv/v3/tv/${ids[i]}?api_key=${fanartTVClientID}`;
    fetch(fanartShowURL)
      .then(response => response.json())
      .then((json) => {
        const posterURL = json.tvposter ? json.tvposter[0].url : 'https://www.classicposters.com/images/nopicture.gif';
        return dispatch(getPosterUrl(posterURL, ids[i]));
      });
  }
};

export function fetchPosterIfNeeded(ids) {
  return (dispatch, getState) => {
    if (shouldFetchPoster(getState())) {
      return dispatch(fetchPoster(ids));
    }
  };
}
