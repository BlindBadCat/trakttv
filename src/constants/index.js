export default {
  FETCH_SHOWS_REQUEST: 'FETCH_SHOWS_REQUEST',
  FETCH_SHOWS_SUCCESS: 'FETCH_SHOWS_SUCCESS',
  FETCH_SHOWS_ERROR: 'FETCH_SHOWS_ERROR',
  CHANGE_URL_PARAMS: 'CHANGE_URL_PARAMS',
};
