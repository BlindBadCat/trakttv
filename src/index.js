import React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import storeFactory from './store/configureStore';
import registerServiceWorker from './registerServiceWorker';
import App from './containers/App';

const store = storeFactory();

ReactDOM.render(
  <Provider store={store}>
    <App />
  </Provider>,
  document.getElementById('root'),
);
registerServiceWorker();
