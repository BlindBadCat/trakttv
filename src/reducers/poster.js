import C from '../constants';

export const initialState = {
  isFetching: false,
  urls: [],
};

const poster = (state = initialState, { type, payload }) => {
  /*switch (type) {
    case C.GET_POSTER_URL_REQUEST:
      return { ...state, isFetching: true };
    case C.GET_POSTER_URL_ADN_ID:
      return { ...state, urls: [...state.urls, payload] };
    case C.RESET_FOR_NEW_FETCH:
      return { ...initialState };
    default: return state;
  }*/
  return state;
};


export default poster;
