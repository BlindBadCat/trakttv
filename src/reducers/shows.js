import C from '../constants';

export const initialState = {
  isFetching: false,
  shows: [],
  sort: 'shows/watched/all',
  query: '',
  genre: '',
  pageCount: 5,
  limit: 10,
  currentPage: 1,
  itemCount: 10,
};

const shows = (state = initialState, { type, payload }) => {
  switch (type) {
    case C.FETCH_SHOWS_REQUEST:
      return { ...state, isFetching: true };
    case C.FETCH_SHOWS_SUCCESS:
      return {
        ...state,
        shows: [...payload.shows],
        ...payload.pagination,
        isFetching: false,
      };
    case C.FETCH_SHOWS_ERROR:
      return { ...state };
    case C.CHANGE_URL_PARAMS:
      return { ...state, ...payload };
    default: return state;
  }
};

export default shows;
