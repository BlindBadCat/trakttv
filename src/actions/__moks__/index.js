import shows from '../samples/shows';

export const fetchPosterIfNeeded = urlParams => {
  Promise.resolve({
    json: shows,
  })
}