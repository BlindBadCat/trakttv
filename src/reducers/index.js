import { combineReducers } from 'redux';
import { routerReducer } from 'react-router-redux';
import shows from './shows';

const rootReducer = combineReducers({
  routerReducer,
  shows,
});

export default rootReducer;
