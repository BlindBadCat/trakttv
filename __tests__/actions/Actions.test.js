import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock from 'fetch-mock';
import C from '../../src/constants';
import * as actions from '../../src/actions';
import * as types from '../../src/constants';
import sample from '../../src/actions/samples/shows';
import { initialState } from '../../src/reducers/shows';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('async actions', () => {
  afterEach(() => {
    fetchMock.reset();
    fetchMock.restore();
  });

  it('fetchs data from traktTV', () => {
    fetchMock.getOnce('*', {...sample});
    const expectedActions = [
      { type: C.FETCH_SHOWS_REQUEST },
      {
        type: C.FETCH_SHOWS_SUCCESS,
        payload: {
          shows: { ...sample },
          pagination: {
            pageCount: 4000,
            limit: 10,
            currentPage: 1,
            itemCount: 40000,
          },
        },
      },
    ];
    const store = mockStore({...initialState});â€‹
    return store.dispatch(actions.fetchShows()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions)
    })
  });
});
