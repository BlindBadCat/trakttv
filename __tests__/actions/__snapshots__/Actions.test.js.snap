// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action test Dispatches the correct action and payload 1`] = `
Array [
  Object {
    "type": "FETCH_SHOWS_REQUEST",
  },
]
`;
